docs/proposals/external-ca.md

```diff
@@ -0,0 +1,112 @@
+# External Certificate Authority Proposal
+
+## Context
+
+To enable mTLS in the mesh, each dataplane has to receive a unique certificate. This certificate has to be signed by the root certificate which is unique for every mesh.
+Right now, Kuma auto-generates root certificate when mesh is created. Root certificates are stored in Postgres on Universal and as a Secret on Kubernetes.
+In many organisations there is already Certificate Authority, so there should be a way to use it with Kuma.
+
+## Requirements
+
+User wants to generate root certificate and provide it to Kuma.
+
+## Proposed configuration model
+
+```yaml
+type: Mesh
+name: default
+mtls:
+  ca:
+    external: {}
+  enabled: true
+```
+
+## How to provide certificates to Kuma
+
+There are several ways we can implement providing root certs for Kuma
+
+### File
+
+```yaml
+type: Mesh
+name: default
+mtls:
+  ca:
+    external:
+      keyPath: /path/to/file
+      certPath: /path/to/file
+  enabled: true
+```
+
+It seems simple, but storing key files on the machines with control plane is not ideal.
+The bigger problem though is that you have to synchronize files between CP instances. Once it's not in sync (Puppet fails, any other reason), the mesh will break.
+We already have storage to share between CP instances and it's Postgres on Universal and Etcd on K8S.
+  
+### HTTP Server + kumactl
+
+We can adopt the same pattern as with Dataplane Token Server.
+The server is by default exposed only on `localhost`. Optionally user can expose server to public by generating cert+key for server and client certificates.
+
+User can upload a certificate with key in two ways.
+
+1) HTTP API
+```bash
+$ cat request.json
+{
+  "key": "-----BEGIN RSA PRIVATE KEY-----\nMLLLEpSDFAGKC24gArqiy1c3pFT3FSk5FE51A4ALAadeR...",
+  "cert": "-----BEGIN CERTIFICATE-----\nZZXDFjC3An4gAwFBAFIJ4FFZ66emAA3AZA0GZSqGSGHt...",
+  "mesh": "demo"
+}
+$ curl -XPUT localhost:5691/certificates --data @request.json
+```
+
+2) kumactl
+```bash
+$ cat key.pem
+-----BEGIN RSA PRIVATE KEY-----\nMLLLEpSDFAGKC24gArqiy1c3pFT3FSk5FE51A4ALAadeR...
+$ cat cert.pem
+-----BEGIN CERTIFICATE-----\nZZXDFjC3An4gAwFBAFIJ4FFZ66emAA3AZA0GZSqGSGHt...
+$ kumactl upload-certificate --mesh demo --key-file key.pem --cert-file cert.pem
+```
+
+This can be either yet another HTTP server in Kuma or we can embed this into Dataplane Token Server and rename the server.
+
+### kuma-cp sub command
+
+Right now, you can run the CP with `kuma-cp run`.
+We can introduce new command:
+
+```bash
+$ cat key.pem
+-----BEGIN RSA PRIVATE KEY-----\nMLLLEpSDFAGKC24gArqiy1c3pFT3FSk5FE51A4ALAadeR...
+$ cat cert.pem
+-----BEGIN CERTIFICATE-----\nZZXDFjC3An4gAwFBAFIJ4FFZ66emAA3AZA0GZSqGSGHt...
+$ kuma-cp upload-certificate --mesh demo --key-file key.pem --cert-file cert.pem
```

Comment:
Let's document pros and cons over  `HTTP(S) Server + kumactl`.

Which option do we prefer ourselves ?

Reply:
Added. Please, feel free to contribute to this list!